sync/sync.go

```diff
@@ -2,80 +2,125 @@ package sync
 		log:          log,
-		quit:         make(chan struct{}),
 	}
 }
 
 // Run starts the Synchronizer, returns an error if the loop is already running
-func (s *Synchronizer) Run() error {
-	if running := atomic.CompareAndSwapUint64(&s.running, 0, 1); !running {
-		return errors.New("synchronizer is already running")
-	}
-	return s.SyncBlocks()
+func (s *Synchronizer) Run(ctx context.Context) error {
+	return s.SyncBlocks(ctx)
 }
 
-// Shutdown attempts to stop the Synchronizer, should block until loop acknowledges the request
-func (s *Synchronizer) Shutdown() error {
-	if stopped := atomic.CompareAndSwapUint64(&s.running, 1, 0); !stopped {
-		return errors.New("synchronizer is stopped")
+func (s *Synchronizer) SyncBlocks(syncCtx context.Context) error {
+	errChan := make(chan ErrSyncFailed)
+	fetchers := stream.New().WithMaxGoroutines(runtime.NumCPU())
+	verifiers := stream.New().WithMaxGoroutines(runtime.NumCPU())
+
+	verifierTask := func(ctx context.Context, block *core.Block, stateUpdate *core.StateUpdate) stream.Callback {
+		err := s.Blockchain.SanityCheckNewHeight(block, stateUpdate)
+		return func() {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				if err != nil {
+					s.log.Warnw("Sanity checks failed", "number", block.Number, "hash", block.Hash.Text(16))
+					errChan <- ErrSyncFailed{block.Number, err}
+					return
+				}
+				err = s.Blockchain.Store(block, stateUpdate)
+				if err != nil {
+					s.log.Warnw("Failed storing Block", "number", block.Number, "hash", block.Hash.Text(16),
+						"err", err.Error())
+					errChan <- ErrSyncFailed{block.Number, err}
+					return
+				}
+
+				s.log.Infow("Stored Block", "number", block.Number, "hash", block.Hash.Text(16),
+					"root", block.GlobalStateRoot.Text(16))
+			}
+		}
+	}
+
+	fetcherTask := func(ctx context.Context, height uint64) stream.Callback {
+		var block *core.Block
+		var stateUpdate *core.StateUpdate
+		var err error
+		for {
+			select {
+			case <-ctx.Done():
+				return func() {}
+			default:
+				block, err = s.StarknetData.BlockByNumber(ctx, height)
+				if err != nil {
+					continue
+				}
+				stateUpdate, err = s.StarknetData.StateUpdate(ctx, height)
+				if err != nil {
+					continue
+				}
+
+				return func() {
+					verifiers.Go(func() stream.Callback { return verifierTask(ctx, block, stateUpdate) })
+				}
+			}
+		}
+	}
+
+	streamCtx, streamCancel := context.WithCancel(syncCtx)
+	nextHeight := uint64(0)
+	if h, err := s.Blockchain.Height(); err == nil {
+		nextHeight = h + 1
 	}
-	close(s.quit)
-	return nil
-}
 
-func (s *Synchronizer) SyncBlocks() error {
 	for {
 		select {
-		case <-s.quit:
-			// Clean up and return
-			return nil
+		case err := <-errChan:
+			streamCancel() // cancel all running tasks
+			streamCtx, streamCancel = context.WithCancel(syncCtx)
+			nextHeight = err.Height // keep syncing from failed height
+		case <-syncCtx.Done():
+			fetchers.Wait()
+			verifiers.Wait()
+			return syncCtx.Err()
```

Comment:
Should we be returning `nil` here? Right now, the node prints an error message when I `<Ctrl-c>`.

Reply:
Maybe we should check for “context canceled” error and return nil in that case. But I am not sure if that is a special type.

Reply:
done